sys/net/gnrc/link_layer/lwmac/lwmac.c

```diff
@@ -0,0 +1,745 @@
+        return;
+
+    if(newstate >= STATE_COUNT) {
+        LOG_ERROR("Trying to set invalid state %u\n", newstate);
+        return;
+    }
+
+    /* Already change state, but might be reverted to oldstate when needed */
+    lwmac.state = newstate;
+
+    /* Actions when leaving old state */
+    switch(oldstate)
+    {
+    case RECEIVING:
+    case TRANSMITTING:
+        /* Enable duty cycling again */
+        rtt_handler(LWMAC_EVENT_RTT_RESUME);
+        break;
+
+    case SLEEPING:
+        lwmac_clear_timeout(&lwmac, TIMEOUT_WAKEUP_PERIOD);
+        break;
+
+    default:
+        break;
+    }
+
+    /* Actions when entering new state */
+    switch(newstate)
+    {
+    /*********************** Operation states *********************************/
+    case LISTENING:
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);
+        break;
+
+    case SLEEPING:
+        /* Put transceiver to sleep */
+        _set_netdev_state(&lwmac, NETOPT_STATE_SLEEP);
+        /* We may have come here through RTT handler, so timeout may still be active */
+        lwmac_clear_timeout(&lwmac, TIMEOUT_WAKEUP_PERIOD);
+        /* Return immediately, so no rescheduling */
+        return;
+
+    /* Trying to send data */
+    case TRANSMITTING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while RXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    /* Receiving incoming data */
+    case RECEIVING:
+        rtt_handler(LWMAC_EVENT_RTT_PAUSE); /* No duty cycling while TXing */
+        _set_netdev_state(&lwmac, NETOPT_STATE_IDLE);  /* Power up netdev */
+        break;
+
+    case STOPPED:
+        _set_netdev_state(&lwmac, NETOPT_STATE_OFF);
+        break;
+
+    /*********************** Control states ***********************************/
+    case START:
+        rtt_handler(LWMAC_EVENT_RTT_START);
+        lwmac_set_state(LISTENING);
+        break;
+
+    case STOP:
+        rtt_handler(LWMAC_EVENT_RTT_STOP);
+        lwmac_set_state(STOPPED);
+        break;
+
+    case RESET:
+        LOG_WARNING("Reset not yet implemented\n");
+        lwmac_set_state(STOP);
+        lwmac_set_state(START);
+        break;
+
+    /**************************************************************************/
+    default:
+        LOG_DEBUG("No actions for entering state %u\n", newstate);
+        return;
+    }
+
+    lwmac_schedule_update();
+}
+
+/******************************************************************************/
+
+/* Main state machine. Call whenever something happens */
+// TODO: Don't use global variables
+bool lwmac_update(void)
+{
+    lwmac.needs_rescheduling = false;
+
+    switch(lwmac.state)
+    {
+    case SLEEPING:
+
+        /* If a packet is scheduled, no other (possible earlier) packet can be
+         * sent before the first one is handled, even no broadcast
+         */
+        if( !lwmac_timeout_is_running(&lwmac, TIMEOUT_WAIT_FOR_DEST_WAKEUP) ) {
```

Comment:
Sorry that I am a little bit fuzzy here. When the device is in the sleeping period, it will be in the standby mode (low-power mode) right? Based on my understanding of the related part of your thesis which I have read, it seems to me that the xtimer module will not work when the device has been put into the standby mode during sleeping (low-power mode). So, is it still meaningful to check running timeouts here (will there be any timeouts)? :-)